iis-wf/iis-wf-referenceextraction/src/main/java/eu/dnetlib/iis/wf/referenceextraction/softwareurl/CachedWebCrawlerJob.java

```diff
@@ -62,156 +59,88 @@ public static void main(String[] args) throws Exception {
-
-            LockManager lockManager = LockManagerUtils.instantiateLockManager(params.lockManagerFactoryClassName,
-                    sc.hadoopConfiguration());
             
-            OutputPaths outputPaths = new OutputPaths(params);
+            Configuration hadoopConf = sc.hadoopConfiguration();
+            
+            HdfsUtils.remove(hadoopConf, params.getOutputPath());
+            HdfsUtils.remove(hadoopConf, params.getOutputFaultPath());
+            HdfsUtils.remove(hadoopConf, params.getOutputReportPath());
+
+            LockManager lockManager = LockManagerUtils.instantiateLockManager(params.getLockManagerFactoryClassName(),
+                    hadoopConf);
             
             JavaRDD<DocumentToSoftwareUrl> documentToSoftwareUrl = avroLoader.loadJavaRDD(sc, params.inputPath,
                     DocumentToSoftwareUrl.class);
             
-            if (documentToSoftwareUrl.isEmpty()) {
-                storeInOutput(sc.emptyRDD(), sc.emptyRDD(),
-                        generateReportEntries(sc, sc.emptyRDD(), sc.emptyRDD(), sc.emptyRDD()), 
-                        outputPaths, contentRetrieverContext.getNumberOfEmittedFiles());
-                return;
-            }
+            final String cacheRootDir = CacheStorageUtils.normalizePath(params.getCacheRootDir());
+            CacheMetadataManagingProcess cacheManager = new CacheMetadataManagingProcess();
+            String existingCacheId = cacheManager.getExistingCacheId(hadoopConf, cacheRootDir);
             
-            // getting existing cache id
-            String existingCacheId = cacheManager.getExistingCacheId(sc.hadoopConfiguration(), cacheRootDir);
+            // skipping already extracted
+            JavaRDD<DocumentText> cachedSources = CacheStorageUtils.getRddOrEmpty(sc, avroLoader, cacheRootDir,
+                    existingCacheId, CacheRecordType.text, DocumentText.class);
+            // will be written in new cache version and output
+            cachedSources.cache();
             
-            // checking whether cache is empty
-            if (CacheMetadataManagingProcess.UNDEFINED.equals(existingCacheId)) {
+            JavaPairRDD<CharSequence, CharSequence> cacheByUrl = cachedSources.mapToPair(x -> new Tuple2<>(x.getId(), x.getText()));
+            JavaPairRDD<CharSequence, DocumentToSoftwareUrl> inputByUrl = documentToSoftwareUrl.mapToPair(x -> new Tuple2<>(x.getSoftwareUrl(), x));
+            JavaPairRDD<CharSequence, Tuple2<DocumentToSoftwareUrl, Optional<CharSequence>>> inputJoinedWithCache = inputByUrl.leftOuterJoin(cacheByUrl);
+
+            JavaRDD<DocumentToSoftwareUrl> toBeProcessed = inputJoinedWithCache.filter(x -> !x._2._2.isPresent()).values().map(x -> x._1);
+            JavaRDD<DocumentToSoftwareUrlWithSource> entitiesReturnedFromCache = inputJoinedWithCache.filter(x -> x._2._2.isPresent()).values().map(x -> attachSource(x._1, x._2.get()));
```

Comment:
OK, it is kind of a separate task but since the `CachedWebCrawlerJob` code was modified in this area I will run more tests to evaluate the caching impact on webcrawer performance.

Reply:
OK, similarly to one of the comments above this change did not have any significant impact on overall performance.

Looking at substages it is clear the stage including writing the cache takes more time than the one without it and similarly (and proportionally) the subsequent stage in the job version with caching takes less while the version without the cache - more time. But the gain is basically equal to the cost.